### ISSUE TITLE
[Yul] optimizer optimizes out side-effects that change the value returned by msize()?

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#7098]
## Description

Found by oss-fuzz: issue 15818

```
{
	let a

	function foo_singlereturn_0() -> out
	{
		mstore(lt(or(gt(1,or(or(gt(or(or(or(1,gt(or(gt(or(or(keccak256(pc(),or(gt(not(pc()),1),1)),1),not(1)),pc()),1),pc())),lt(or(1,sub(pc(),1)),1)),pc()),1),1),gt(not(pc()),1))),1),1),1)
		sstore(not(pc()),1)
	}

	function foo_singlereturn_1(in_1, in_2) -> out
	{
		extcodecopy(1,msize(),1,1)
	}

	a := foo_singlereturn_0()
	sstore(0,0)
	sstore(2,1)

	a := foo_singlereturn_1(calldataload(0),calldataload(3))
	sstore(0,0)
	sstore(3,1)
}
```

produces the following trace pre and post optimization

```
// Pre optimization trace
  EXTCODECOPY(1, 128, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    80: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

```
// Post optimization trace
  EXTCODECOPY(1, 32, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    20: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): petersburg

## Steps to Reproduce

To obtain pre-optimization trace, do

```
$ yulrun --input-file <code.yul> run
Trace:
  EXTCODECOPY(1, 128, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    80: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

To obtain post-optimized trace, first obtain optimized yul code like so
```
$ solc --strict-assembly --optimize <code.yul> 2>/dev/null | | grep -Pzo "code {\n(.|\n)*}\n\n"
code {
        {
            let _1 := gt(not(pc()), 1)
            let _2 := pc()
            let _3 := not(0)
            let _4 := lt(or(1, add(pc(), _3)), 1)
            let _5 := pc()
            let _6 := pc()
            mstore(lt(or(gt(1, or(or(gt(or(or(or(gt(or(gt(_3, _6), 1), _5), _4), _2), 1), 1), _1), 1)), 1), 1), 1)
            sstore(not(pc()), 1)
            sstore(0, 0)
            sstore(2, 1)
            pop(foo_singlereturn_1(calldataload(0), calldataload(3)))
            sstore(0, 0)
            sstore(3, 1)
        }
        function foo_singlereturn_1(in, in_1) -> out
        { extcodecopy(1, msize(), 1, 1) }
    }
}
```
Then run yulrun against the optimized yul code
```
$ yulrun --input-file <optimized_yul_snippet_shown_above.yul> run
Trace:
  EXTCODECOPY(1, 32, 1, 1)
Memory dump:
     0: 0000000000000000000000000000000000000000000000000000000000000001
    20: 6f00000000000000000000000000000000000000000000000000000000000000
Storage dump:
  0000000000000000000000000000000000000000000000000000000000000002: 0000000000000000000000000000000000000000000000000000000000000001
  0000000000000000000000000000000000000000000000000000000000000003: 0000000000000000000000000000000000000000000000000000000000000001
  ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff88: 0000000000000000000000000000000000000000000000000000000000000001
```

**Comment by bshastry** [comment]
Actually, #6830 makes an `msize()` call in the yul interpreter return a constant. So I'm not really sure what's happening.

**Comment by bshastry** [comment]
Oh, iiuc #6830 was not merged because we decided to fix a similar bug earlier #6827 via #6840

**Comment by bshastry** [comment]
I can reproduce this issue on `>=0.5.9`
